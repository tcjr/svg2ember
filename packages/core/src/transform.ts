import { parse, type SvgNode } from 'svg-parser';
import { optimize } from 'svgo';
import type { TransformOptions, TransformResult } from './types.js';

export function transform(
  svgContent: string,
  options: TransformOptions = {},
): TransformResult {
  const {
    typescript = false,
    optimize: shouldOptimize = true,
    svgoConfig = {},
  } = options;

  let processedSvg = svgContent;

  // Optimize SVG if requested
  if (shouldOptimize) {
    const result = optimize(svgContent, {
      multipass: true,
      plugins: [
        {
          name: 'preset-default',
          params: {
            overrides: {
              removeViewBox: false,
            },
          },
        },
        'removeXMLNS',
      ],
      ...svgoConfig,
    });
    processedSvg = result.data;
  }

  // Parse SVG using svg-parser for AST-based processing
  let ast;
  try {
    ast = parse(processedSvg);
  } catch {
    throw new Error('Invalid SVG: Failed to parse SVG content');
  }

  if (!ast.children || ast.children.length === 0) {
    throw new Error('Invalid SVG: No content found');
  }

  // Find the root SVG element
  const svgElement = ast.children.find(
    (child: SvgNode) => child.type === 'element' && child.tagName === 'svg',
  );

  if (!svgElement) {
    throw new Error('Invalid SVG: No <svg> root element found');
  }

  // Generate Ember component template
  const componentCode = generateEmberComponent(svgElement, typescript);
  const extension = typescript ? '.gts' : '.gjs';

  return {
    code: componentCode,
    extension,
  };
}

function generateEmberComponent(
  svgElement: SvgNode,
  typescript: boolean,
): string {
  // Add a marker for ...attributes to the root svg element's properties
  // SvgNode properties are optional, so ensure it exists.
  // svgElement is guaranteed to be an 'element' type here due to prior checks.
  if (!svgElement.properties) {
    svgElement.properties = {};
  }
  svgElement.properties['__spreadAttributes__'] = 1;

  // Convert AST back to SVG string; astToSvgString will handle the spread
  const svgWithAttributes = astToSvgString(svgElement);

  if (typescript) {
    return `import type { TOC } from '@ember/component/template-only';

interface Signature {
  Element: SVGSVGElement;
}

const SvgComponent: TOC<Signature> = <template>
  ${svgWithAttributes}
</template>;

export default SvgComponent;`;
  } else {
    return `<template>
  ${svgWithAttributes}
</template>`;
  }
}

function astToSvgString(element: SvgNode): string {
  if (element.type !== 'element') {
    return element.type === 'text' ? element.value || '' : '';
  }

  const { tagName, properties = {}, children = [] } = element;

  // Convert properties to attribute string
  const attrs = Object.entries(properties)
    .map(([key, value]) => {
      if (key === '__spreadAttributes__' && value === 1) {
        return '...attributes';
      }
      return `${key}="${value}"`;
    })
    .join(' ');

  const attrsString = attrs ? ` ${attrs}` : '';

  if (children.length === 0) {
    return `<${tagName}${attrsString} />`;
  }

  const childrenString = children
    .map((child: SvgNode) => astToSvgString(child))
    .join('');

  return `<${tagName}${attrsString}>${childrenString}</${tagName}>`;
}
