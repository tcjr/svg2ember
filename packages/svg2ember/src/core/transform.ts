import { parse, type SvgNode } from 'svg-parser';
import { optimize, type Config } from 'svgo';
import type { TransformOptions, TransformResult } from './types.js';

export const DEFAULT_SVGO_CONFIG: Config = {
  multipass: true,
  plugins: [
    {
      name: 'preset-default',
      params: {
        overrides: {
          removeViewBox: false,
        },
      },
    },
    'removeXMLNS',
  ],
};

export function transform(
  svgContent: string,
  options: TransformOptions = {},
): TransformResult {
  const {
    typescript = false,
    optimize: shouldOptimize = true,
    svgoConfig = DEFAULT_SVGO_CONFIG,
  } = options;

  let processedSvg = svgContent;

  // Optimize SVG if requested
  if (shouldOptimize) {
    const result = optimize(svgContent, svgoConfig);
    processedSvg = result.data;
  }

  // Parse SVG using svg-parser for AST-based processing
  let ast;
  try {
    ast = parse(processedSvg);
  } catch {
    throw new Error('Invalid SVG: Failed to parse SVG content');
  }

  if (!ast.children || ast.children.length === 0) {
    throw new Error('Invalid SVG: No content found');
  }

  // Find the root SVG element
  const svgElement = ast.children.find(
    (child: SvgNode) => child.type === 'element' && child.tagName === 'svg',
  );

  if (!svgElement) {
    throw new Error('Invalid SVG: No <svg> root element found');
  }

  // Generate Ember component template
  const componentCode = generateEmberComponent(svgElement, typescript);
  const extension = typescript ? '.gts' : '.gjs';

  return {
    code: componentCode,
    extension,
  };
}

function generateEmberComponent(
  svgElement: SvgNode,
  typescript: boolean,
): string {
  // Convert AST back to SVG string with attributes spread
  const svgString = astToSvgString(svgElement);

  if (typescript) {
    return `import type { TOC } from '@ember/component/template-only';

interface Signature {
  Element: SVGSVGElement;
}

const IconComponent: TOC<Signature> = <template>
  ${svgString}
</template>;

export default IconComponent;`;
  } else {
    return `<template>
  ${svgString}
</template>`;
  }
}

function astToSvgString(element: SvgNode): string {
  if (element.type !== 'element') {
    return element.type === 'text' ? element.value || '' : '';
  }

  const { tagName, properties = {}, children = [] } = element;

  // Convert properties to attribute string
  const attrs = Object.entries(properties)
    .map(([key, value]) => `${key}="${value}"`)
    .join(' ');

  let attrsString = attrs ? ` ${attrs}` : '';

  // Never self-close <svg> because we add {{yield}}
  if (children.length === 0 && tagName !== 'svg') {
    return `<${tagName}${attrsString} />`;
  }

  let childrenString = children
    .map((child: SvgNode) => astToSvgString(child))
    .join('');

  if (tagName === 'svg') {
    // Add `...attributes` to the opening `<svg>` tag
    attrsString = `${attrsString} ...attributes`;

    // Add `{{yield}}` just before the closing `</svg>`
    childrenString = `${childrenString}{{yield}}`;
  }

  return `<${tagName}${attrsString}>${childrenString}</${tagName}>`;
}
