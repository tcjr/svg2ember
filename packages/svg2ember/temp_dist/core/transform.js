"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SVGO_CONFIG = void 0;
exports.transform = transform;
var svg_parser_1 = require("svg-parser");
var svgo_1 = require("svgo");
exports.DEFAULT_SVGO_CONFIG = {
    multipass: true,
    plugins: [
        {
            name: 'preset-default',
            params: {
                overrides: {
                    removeViewBox: false,
                },
            },
        },
        'removeXMLNS',
    ],
};
function transform(svgContent, options) {
    if (options === void 0) { options = {}; }
    var _a = options.typescript, typescript = _a === void 0 ? false : _a, _b = options.optimize, shouldOptimize = _b === void 0 ? true : _b, _c = options.svgoConfig, svgoConfig = _c === void 0 ? exports.DEFAULT_SVGO_CONFIG : _c;
    var processedSvg = svgContent;
    // Optimize SVG if requested
    if (shouldOptimize) {
        var result = (0, svgo_1.optimize)(svgContent, svgoConfig);
        processedSvg = result.data;
    }
    // Parse SVG using svg-parser for AST-based processing
    var ast;
    try {
        ast = (0, svg_parser_1.parse)(processedSvg);
    }
    catch (_d) {
        throw new Error('Invalid SVG: Failed to parse SVG content');
    }
    if (!ast.children || ast.children.length === 0) {
        throw new Error('Invalid SVG: No content found');
    }
    // Find the root SVG element
    var svgElement = ast.children.find(function (child) { return child.type === 'element' && child.tagName === 'svg'; });
    if (!svgElement) {
        throw new Error('Invalid SVG: No <svg> root element found');
    }
    // Generate Ember component template
    var componentCode = generateEmberComponent(svgElement, typescript);
    var extension = typescript ? '.gts' : '.gjs';
    return {
        code: componentCode,
        extension: extension,
    };
}
function generateEmberComponent(svgElement, typescript) {
    // Convert AST back to SVG string with attributes spread
    var svgString = astToSvgString(svgElement);
    if (typescript) {
        return "import type { TOC } from '@ember/component/template-only';\n\ninterface Signature {\n  Element: SVGSVGElement;\n}\n\nconst IconComponent: TOC<Signature> = <template>\n  ".concat(svgString, "\n</template>;\n\nexport default IconComponent;");
    }
    else {
        return "<template>\n  ".concat(svgString, "\n</template>");
    }
}
function astToSvgString(element) {
    if (element.type !== 'element') {
        return element.type === 'text' ? element.value || '' : '';
    }
    var tagName = element.tagName, _a = element.properties, properties = _a === void 0 ? {} : _a, _b = element.children, children = _b === void 0 ? [] : _b;
    // Convert properties to attribute string
    var attrs = Object.entries(properties)
        .map(function (_a) {
        var key = _a[0], value = _a[1];
        return "".concat(key, "=\"").concat(value, "\"");
    })
        .join(' ');
    var attrsString = attrs ? " ".concat(attrs) : '';
    // Never self-close <svg> because we add {{yield}}
    if (children.length === 0 && tagName !== 'svg') {
        return "<".concat(tagName).concat(attrsString, " />");
    }
    var childrenString = children
        .map(function (child) { return astToSvgString(child); })
        .join('');
    if (tagName === 'svg') {
        // Add `...attributes` to the opening `<svg>` tag
        attrsString = "".concat(attrsString, " ...attributes");
        // Add `{{yield}}` just before the closing `</svg>`
        childrenString = "".concat(childrenString, "{{yield}}");
    }
    return "<".concat(tagName).concat(attrsString, ">").concat(childrenString, "</").concat(tagName, ">");
}
