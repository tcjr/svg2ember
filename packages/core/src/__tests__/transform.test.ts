import { describe, it, expect } from 'vitest';
import { transform } from '../transform.js';

describe('transform', () => {
  const basicSvg = `<svg width="24" height="24" viewBox="0 0 24 24">
  <path d="M12 2L2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7l-10-5z"/>
</svg>`;

  it('should transform basic SVG to JavaScript component', () => {
    const result = transform(basicSvg, { typescript: false });

    expect(result.extension).toBe('.gjs');
    expect(result.code).toMatchSnapshot();
  });

  it('should transform basic SVG to TypeScript component', () => {
    const result = transform(basicSvg, { typescript: true });

    expect(result.extension).toBe('.gts');
    expect(result.code).toMatchSnapshot();
  });

  it('should add attributes spread to svg element', () => {
    const result = transform(basicSvg);

    expect(result.code).toMatch(/<svg[^>]*\s+\.\.\.attributes[^>]*>/);
    expect(result.code).toMatchSnapshot();
  });

  it('should preserve SVG attributes', () => {
    const result = transform(basicSvg);

    expect(result.code).toContain('width="24"');
    expect(result.code).toContain('height="24"');
    expect(result.code).toContain('viewBox="0 0 24 24"');
    expect(result.code).toMatchSnapshot();
  });

  it('should handle SVG optimization', () => {
    const svgWithId = `
<svg width="24" height="24" viewBox="0 0 24 24" id="SVG_ID_1">
  <path d="M12 2L2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7l-10-5z"/>
</svg>`;

    const result = transform(svgWithId);

    // Ids should be removed by optimization
    expect(result.code).not.toContain('SVG_ID_1');
    expect(result.code).toMatchSnapshot();
  });

  it('should handle SVG with no optimization', () => {
    const svgWithId = `
<svg width="24" height="24" viewBox="0 0 24 24" id="SVG_ID_1">
  <path d="M12 2L2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7l-10-5z"/>
</svg>`;
    const result = transform(svgWithId, { optimize: false });
    expect(result.code).toContain('id="SVG_ID_1"');
    expect(result.code).toMatchSnapshot();
  });

  it('should throw error for invalid SVG', () => {
    expect(() => transform('not an svg')).toThrow();
  });

  it('should throw error for SVG with no root element', () => {
    expect(() => transform('<div>not svg</div>')).toThrow(
      'Invalid SVG: No <svg> root element found',
    );
  });

  it('should handle complex SVG with nested elements', () => {
    const complexSvg = `
<?xml version="1.0" encoding="UTF-8"?>
<svg width="88px" height="88px" viewBox="0 0 88 88" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 46.2 (44496) - http://www.bohemiancoding.com/sketch -->
    <title>Dismiss</title>
    <desc>Created with Sketch.</desc>
    <defs></defs>
    <g id="Blocks" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="square">
        <g id="Dismiss" stroke="#063855" stroke-width="2">
            <path d="M51,37 L37,51" id="Shape"></path>
            <path d="M51,51 L37,37" id="Shape"></path>
        </g>
    </g>
</svg>
`;

    const result = transform(complexSvg, { optimize: false });
    expect(result.code).toMatchSnapshot();
  });

  it('should transform empty SVG', () => {
    const result = transform('<svg></svg>');
    expect(result.extension).toBe('.gjs');
    expect(result.code).toMatchSnapshot();
  });

  it('should transform empty self-closing SVG', () => {
    const result = transform('<svg />');
    expect(result.extension).toBe('.gjs');
    expect(result.code).toMatchSnapshot();
  });
});

describe('transform with svgoConfig', () => {
  it('should pass options onto svgo, with xmlns', () => {
    const svgSource = `
      <svg xmlns="http://www.w3.org/2000/svg">
      </svg>
    `;
    const result = transform(svgSource, {
      svgoConfig: {
        plugins: [
          {
            name: 'preset-default',
            params: {
              overrides: {
                removeViewBox: false,
              },
            },
          },
          //'removeXMLNS', <- commenting this means xmlns will remain
        ],
      },
    });
    expect(result.code).toContain('xmlns');
  });

  it('should pass options onto svgo, without xmlns', () => {
    const svgSource = `
      <svg xmlns="http://www.w3.org/2000/svg">
      </svg>
    `;
    const result = transform(svgSource, {
      svgoConfig: {
        plugins: [
          {
            name: 'preset-default',
            params: {
              overrides: {
                removeViewBox: false,
              },
            },
          },
          'removeXMLNS', // <- leaving this means xmlns will be removed
        ],
      },
    });
    expect(result.code).not.toContain('xmlns');
  });
});
