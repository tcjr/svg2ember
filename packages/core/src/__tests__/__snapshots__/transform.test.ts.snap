// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > should add attributes spread to svg element 1`] = `
"<template>
  <svg width="24" height="24" viewBox="0 0 24 24" ...attributes><path d="M12 2 2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7L14 2z" />{{yield}}</svg>
</template>"
`;

exports[`transform > should handle SVG optimization 1`] = `
"<template>
  <svg width="24" height="24" viewBox="0 0 24 24" ...attributes><path d="M12 2 2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7L14 2z" />{{yield}}</svg>
</template>"
`;

exports[`transform > should handle SVG with no optimization 1`] = `
"<template>
  <svg width="24" height="24" viewBox="0 0 24 24" id="SVG_ID_1" ...attributes><path d="M12 2L2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7l-10-5z" />{{yield}}</svg>
</template>"
`;

exports[`transform > should handle complex SVG with nested elements 1`] = `
"<template>
  <svg width="88px" height="88px" viewBox="0 0 88 88" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" ...attributes><title>Dismiss</title><desc>Created with Sketch.</desc><defs /><g id="Blocks" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="square"><g id="Dismiss" stroke="#063855" stroke-width="2"><path d="M51,37 L37,51" id="Shape" /><path d="M51,51 L37,37" id="Shape" /></g></g>{{yield}}</svg>
</template>"
`;

exports[`transform > should preserve SVG attributes 1`] = `
"<template>
  <svg width="24" height="24" viewBox="0 0 24 24" ...attributes><path d="M12 2 2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7L14 2z" />{{yield}}</svg>
</template>"
`;

exports[`transform > should transform basic SVG to JavaScript component 1`] = `
"<template>
  <svg width="24" height="24" viewBox="0 0 24 24" ...attributes><path d="M12 2 2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7L14 2z" />{{yield}}</svg>
</template>"
`;

exports[`transform > should transform basic SVG to TypeScript component 1`] = `
"import type { TOC } from '@ember/component/template-only';

interface Signature {
  Element: SVGSVGElement;
}

const IconComponent: TOC<Signature> = <template>
  <svg width="24" height="24" viewBox="0 0 24 24" ...attributes><path d="M12 2 2 7v10c0 5.55 3.84 10 9 11 1.08.2 2.92.2 4 0 5.16-1 9-5.45 9-11V7L14 2z" />{{yield}}</svg>
</template>;

export default IconComponent;"
`;

exports[`transform > should transform empty SVG 1`] = `
"<template>
  <svg ...attributes>{{yield}}</svg>
</template>"
`;

exports[`transform > should transform empty self-closing SVG 1`] = `
"<template>
  <svg ...attributes>{{yield}}</svg>
</template>"
`;
